lib/kernel/src/application.erl

```diff
@@ -160,135 +169,9 @@ ensure_all_started(Application, Type) ->
 ensure_all_started(Application, Type, Mode) when is_atom(Application) ->
     ensure_all_started([Application], Type, Mode);
 ensure_all_started(Applications, Type, Mode) when is_list(Applications) ->
-    Opts = #{type => Type, mode => Mode},
-
-    case enqueue_or_start(Applications, [], #{}, [], [], Opts) of
-        {ok, DAG, _Pending, Started} when Mode =:= concurrent ->
-            ReqIDs = gen_server:reqids_new(),
-            concurrent_dag_start(maps:to_list(DAG), ReqIDs, [], Started, Type);
-        {ok, DAG, _Pending, Started} when Mode =:= serial ->
-            0 = map_size(DAG),
-            {ok, lists:reverse(Started)};
-        {error, AppReason, Started} ->
-            _ = [stop(Name) || Name <- Started],
-            {error, AppReason}
-    end.
-
-enqueue_or_start([App | Apps], Optional, DAG, Pending, Started, Opts)
-  when is_map_key(App, DAG) ->
-    %% We already traversed the application, so only add it as pending
-    enqueue_or_start(Apps, Optional, DAG, [App | Pending], Started, Opts);
-
-enqueue_or_start([App | Apps], Optional, DAG, Pending, Started, Opts) when is_atom(App) ->
-    %% In case the app is already running, we just skip it instead
-    %% of attempting to start all of its children - which would
-    %% have already been loaded and started anyway.
-    case application_controller:is_running(App) of
```

Comment:
The new code is removing this check, which is a very important optimization: https://github.com/erlang/otp/commit/27f61e9783773fc093bf9a77427d184f474385b6

I am ultimately worried that merging both start and stop paths will remove the option to perform such optimizations. It also makes both paths slower, because we are gathering information from both.

What if you keep the building of the DAG separated (so we no longer need to track parent/child, only the direction relationship) but share the execution of the DAG?

Reply:
Your right, that's a good point, let me split the graph generation for the individual use-cases (we throw away currently one half of the graphs anyways) and put this check back in.